### ISSUE TITLE
Vim9: cannot use member of unknown type

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#7009]
**Describe the bug**

In Vim9 script, we cannot use a member of a dictionary stored in a list of unknown type.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        var ld = []
        ld += [{key: 'value'}]
        def Func()
            eval ld[0].key
        enddef
        Func()
    EOF
    )

`E715` is raised:

    E715: Dictionary required

**Expected behavior**

No error is raised, because:

   1. at the time `ld` was created, it was assigned `[]`
   2. since the list is empty, Vim cannot infer the type of its items
   3. since Vim cannot infer the type of the items, it should infer the list type `list<any>`
   4. if we specify the type `list<any>` when declaring `ld`, no error is raised

```vim
vim9script
var ld: list<any> = []
ld += [{key: 'value'}]
def Func()
    eval ld[0].key
enddef
Func()
```
**Environment**

 - Vim version: 8.2 Included patches: 1-2123
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(362)

**Additional context**

The issue disappears if we specify the type of `ld`:
```vim
vim9script
var ld: list<dict<string>> = []
ld += [{key: 'value'}]
def Func()
    eval ld[0].key
enddef
Func()
```
Or if we evaluate the expression at the script level rather than from a `:def` function:
```vim
vim9script
var ld = []
ld += [{key: 'value'}]
eval ld[0].key
```
---

This issue looks very similar to #6997.

**Comment by lacygoill** [comment]
I encountered this issue in another plugin.  I reduced the code to this:
```vim
vim9
var d = {}
def Func()
    d.key1.key2
enddef
defcompile
```
    E715: Dictionary required

Again, the workaround is to properly declare the type of the variable:
```vim
vim9
var d: dict<dict<any>> = {}
def Func()
    d.key1.key2
enddef
defcompile
```
    ✔

---

Whether this issue is fixed or not, I think we should document the fact that relying on type inferrence for composite expressions might be a bad idea (at least for an empty list/dictionary).

This is OK:
```vim
var n = 0
```
Because `0` is a scalar.

This is NOT OK:
```vim
var d = {}
```
And neither is this:
```vim
var l = []
```
Because `{}` and `[]` are composite expressions; they can contain other types of expressions.

First, it triggers the current issue; but beyond that, writing `= {}` is like writing `: dict<any> = {}`, which might be wrong.  You might want/need to be more specific for your code to work, or to be properly optimized, or for type checking to correctly catch obvious mistakes at compile time.

I can't find a good example where writing `= {}` causes an issue (not a bug, just an expected issue).  When I'll find one, I'll try to send a patch to update the help.

Even a non-empty list/dictionary might give what *seems* to be unexpected errors:
```vim
vim9
var d = {n: 1}
extend(d, {s: 'string'})
```
    E1012: Type mismatch; expected number but got string

That's because `d` was inferred with `dict<number>`.  I find this obvious now, but it wasn't at first.

Bottom line: in a declaration, omitting the type of a scalar is OK; omitting the type of a composite expression is not.
